### PR TITLE
Passkey Credential Creation for Safe Ownership

### LINKED ISSUE #274: [USER STORY] Passkey Credential Creation for Safe Ownership
> As a user, I want to create a new passkey credential so that I can deploy a Safe account owned by this credential

> As a user, I want to be able to sign transactions using my passkey credential to send transactions 

This issue capture the above user stories.

The expected outcome is to have an E2E test that demonstrates:

1. That a signer for a passkey credential can be deployed at a deterministic address
2. That it can function as an owner of a Safe (in the simplest terms, it can be used for a successful call to `execTransaction`)

### PR DESCRIPTION
**Pull request by remedcu**
This PR verifies the user's ability to execute a TX with Safe using that user’s passkey credential.

Closes #274

### CODE REVIEW

File: modules/passkey/test/utils/safe.ts [hunk 1]
```diff
@@ -0,0 +1,45 @@
+import { Hex } from '@noble/curves/abstract/utils'
```
Comment (mmv08) [inline modules/passkey/test/utils/safe.ts:1]: Can this be imported from the 4337 package?

File: modules/passkey/test/userStories/passkeyCredentialCreation.spec.ts [hunk 2]
```diff
@@ -0,0 +1,147 @@
...
+    const { user, safe, safeAddress, signer, navigator, credential } = await setupTests()
+
+    // Send 1 ETH to the Safe
+    await user.sendTransaction({ to: safe, value: ethers.parseEther('1') })
+
+    // Define transaction to be signed.
+    const nonce = await safe.nonce()
+    const randomAddress = ethers.getAddress(ethers.hexlify(ethers.randomBytes(20)))
+    const safeTx = buildSafeTransaction({ to: randomAddress, value: 1n, nonce: nonce })
+    const safeTxData = await safe.encodeTransactionData.staticCall(
```
Comment (mmv08) [inline modules/passkey/test/userStories/passkeyCredentialCreation.spec.ts:91]: Why on-chain? The general rule of thumb is to do things off-chain to save the RPC from some load.

File: modules/passkey/test/userStories/passkeyCredentialCreation.spec.ts [hunk 3]
```diff
@@ -0,0 +1,147 @@
...
+      navigator,
+      credential,
+    }
+  })
+
+  it('should be possible to execute a transaction signed by passkey', async () => {
+    const { user, safe, safeAddress, signer, navigator, credential } = await setupTests()
+
+    // Send 1 ETH to the Safe
+    await user.sendTransaction({ to: safe, value: ethers.parseEther('1') })
```
Comment (mmv08) [inline modules/passkey/test/userStories/passkeyCredentialCreation.spec.ts:85]: does safe really need 1 ether to send 1 wei?

**Comment by nlordell** [comment]
I noticed that we had 2 user story directories and that the file naming was inconsistent. I fixed it here: https://github.com/safe-global/safe-modules/pull/363 , but this PR does not follow the "standard" either, can you fix it as well?

File: modules/passkey/test/utils/safe.ts (hunk 1 above)
Comment (remedcu) [reply modules/passkey/test/utils/safe.ts:1]: Can you specify the path?

File: modules/passkey/test/userStories/passkeyCredentialCreation.spec.ts (hunk 3 above)
Comment (remedcu) [reply modules/passkey/test/userStories/passkeyCredentialCreation.spec.ts]: Not really 😅 Changed 👍🏾

File: modules/passkey/test/utils/safe.ts [hunk 4]
```diff
@@ -0,0 +1,75 @@
...
+import { ethers } from 'hardhat'
+import { Address } from 'hardhat-deploy/types'
+
+const SAFE_TX_TYPEHASH = ethers.keccak256(
+  ethers.toUtf8Bytes(
+    'SafeTx(address to,uint256 value,bytes data,uint8 operation,uint256 safeTxGas,uint256 baseGas,uint256 gasPrice,address gasToken,address refundReceiver,uint256 nonce)',
+  ),
+)
+
+export const domainSeparatorTypehash = ethers.keccak256(ethers.toUtf8Bytes('EIP712Domain(uint256 chainId,address verifyingContract)'))
```
Comment (akshay-ap) [inline modules/passkey/test/utils/safe.ts:11]:
```suggestion
export const DOMAIN_SEPARATOR_TYPEHASH = ethers.keccak256(ethers.toUtf8Bytes('EIP712Domain(uint256 chainId,address verifyingContract)'))
```

Comment (remedcu) [reply modules/passkey/test/utils/safe.ts]: Nice catch! Done 👍🏾

File: modules/passkey/test/utils/safe.ts (hunk 1 above)
Comment (nlordell) [reply modules/passkey/test/utils/safe.ts:1]: I think he means the whole test/utils/safe.ts file.

You should be able to import it from:

```ts
import { buildSafeTransaction, buildSafeTransactionData, DOMAIN_SEPARATOR_TYPEHASH } from '@safe-global/safe-4337/test/utils/safe' 
```

Comment (nlordell) [reply modules/passkey/test/utils/safe.ts:1]: Note that that file is specific to user operations, while here we are interacting with the Safe _without_ 4337. I don't think the aforementioned module contains the code that is used here, but that is rather included in the `@safe-global/safe-contracts` package (which unfortunately targets the wrong `ethers` version).

My intuition is that this module is needed, but I didn't dig too deeply.

File: modules/passkey/test/utils/safe.ts [hunk 5]
```diff
@@ -0,0 +1,75 @@
...
+        safeTx.baseGas,
+        safeTx.gasPrice,
+        safeTx.gasToken,
+        safeTx.refundReceiver,
+        safeTx.nonce,
+      ],
+    ),
+  )
+  return ethers.solidityPacked(['bytes1', 'bytes1', 'bytes32', 'bytes32'], ['0x19', '0x01', domainSeparator, safeTxHash])
+}
```
Comment (nlordell) [inline modules/passkey/test/utils/safe.ts:33]: nit: this should be implemented with the `ethers.TypedDataEncoder` type, which abstracts a lot of this from you already:

File: modules/passkey/test/utils/safe.ts [hunk 6]
```diff
@@ -0,0 +1,75 @@
+import { Hex } from '@noble/curves/abstract/utils'
+import { ethers } from 'hardhat'
+import { Address } from 'hardhat-deploy/types'
```
Comment (nlordell) [inline modules/passkey/test/utils/safe.ts:3]:
```suggestion
import { AddressLike, BytesLike } from 'ethers'
import { ethers } from 'hardhat'
```

File: modules/passkey/test/utils/safe.ts (hunk 5 above)
Comment (nlordell) [reply modules/passkey/test/utils/safe.ts:33]:
```suggestion
export type SafeDomain = {
  verifyingContract: AddressLike
  chainId: BigNumberish
}

export const buildSafeTransactionData = (domain: SafeDomain, safeTx: SafeTransaction): BytesLike => {
  return ethers.TypedDataEncoder.encode(
    domain,
    {
      SafeTx: [
        { name: "to", type: "address"},
        { name: "value", type: "uint256"},
        { name: "data", type: "bytes"},
        { name: "operation", type: "uint8"},
        { name: "safeTxGas", type: "uint256"},
        { name: "baseGas", type: "uint256"},
        { name: "gasPrice", type: "uint256"},
        { name: "gasToken", type: "address"},
        { name: "refundReceiver", type: "address"},
        { name: "nonce", type: "uint256"},
      ]
    },
    safeTx
  )
}
```

File: modules/passkey/test/userstories/PasskeyCredentialCreation.spec.ts [hunk 7]
```diff
@@ -0,0 +1,142 @@
...
+      ethers.ZeroAddress,
+    ])
+    const safeAddress = await proxyFactory.createProxyWithNonce.staticCall(singletonAddress, setupData, 0)
+    await proxyFactory.createProxyWithNonce(singletonAddress, setupData, 0)
+    const safe = await ethers.getContractAt([...SafeL2.abi, ...CompatibilityFallbackHandler.abi], safeAddress)
+
+    const { chainId } = await ethers.provider.getNetwork()
+    const domainSeparator = ethers.keccak256(
+      ethers.AbiCoder.defaultAbiCoder().encode(['bytes32', 'uint256', 'address'], [DOMAIN_SEPARATOR_TYPEHASH, chainId, safeAddress]),
+    )
```
Comment (nlordell) [inline modules/passkey/test/userstories/PasskeyCredentialCreation.spec.ts:74]: We should use `ethers` utilities for this IMO.

File: modules/passkey/test/utils/safe.ts (hunk 6 above)
Comment (remedcu) [reply modules/passkey/test/utils/safe.ts]: `AddressLike` cannot be used for `verifyingContract`, so using `Address` for that, rest changed 👍🏾

File: modules/passkey/test/userstories/PasskeyCredentialCreation.spec.ts (hunk 7 above)
Comment (remedcu) [reply modules/passkey/test/userstories/PasskeyCredentialCreation.spec.ts]: Resolved based on previous comment.

**Review by mmv08 — APPROVED** [review]

**Pull request merged**